Problem: block overrides is not supported in debug_traceCall

mmsqe opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

mmsqe committed:
cleanup proto

mmsqe committed:
add block overrides

mmsqe committed:
Problem: block overrides is not supported in debug_traceCall

mmsqe committed:
add execTrace

mmsqe committed:
cleanup

XinyuCRO commented on rpc/backend/tracing.go:
```diff
@@ -135,6 +135,7 @@ func (b *Backend) convertConfig(config *rpctypes.TraceConfig) *evmtypes.TraceCon
 	}
 	cfg := config.TraceConfig
 	cfg.TracerJsonConfig = string(config.TracerConfig)
+	cfg.BlockOverrides = config.BlockOverrides
```
already signed in `cfg := TraceConfig`
```suggestion
```

mmsqe committed:
check in generic

XinyuCRO reviewed (approved):
LGTM

mmsqe committed:
Update rpc/backend/tracing.go

Co-authored-by: Xinyu <91446598+XinyuCRO@users.noreply.github.com>
Signed-off-by: mmsqe <mavis@crypto.com>

yihuang reviewed (commented):
https://github.com/crypto-org-chain/ethermint/pull/393
let's cleanup ApplyMessageWithConfig api first

mmsqe committed:
Merge remote-tracking branch 'origin/develop' into block_overrides

mmsqe committed:
rm extra tracer

mmsqe committed:
rm traceFn

mmsqe committed:
fix resolve

mmsqe committed:
Merge remote-tracking branch 'origin/develop' into block_overrides

yihuang commented on proto/ethermint/evm/v1/trace_config.proto:
```diff
@@ -36,4 +57,8 @@ message TraceConfig {
+  // temporary state modifications to Geth in order to simulate the effects of eth_call
+  map<string, OverrideAccount> state_overrides = 14 [(gogoproto.jsontag) = "stateOverrides"];
+  // overrides block context fields
+  BlockOverrides block_overrides = 15 [(gogoproto.jsontag) = "blockOverrides"];
```
can we pass json `bytes` to grpc directly, so we only need to decode it once.
and in the json-rpc side, pass it around as `json.RawMessage` like #394

mmsqe committed:
fasten test

mmsqe committed:
update proto

mmsqe committed:
add json raw msg

mmsqe committed:
apply state

mmsqe committed:
test staticcall

yihuang commented on x/evm/keeper/state_transition.go:
```diff
@@ -355,13 +367,20 @@ func (k *Keeper) ApplyMessageWithConfig(
 			return nil, errorsmod.Wrap(err, "failed to apply state override")
 		}
 	}
+	if cfg.DebugTrace && cfg.BlockOverrides != nil {
+		blockCtx := k.NewEVMBlockContext(ctx, cfg)
+		cfg.BlockOverrides.Apply(&blockCtx)
+		evm = k.NewEVMWithBlockCtx(ctx, msg, cfg, stateDB, blockCtx)
+	} else {
+		evm = k.NewEVM(ctx, msg, cfg, stateDB)
+	}
```
```suggestion
	blockCtx := k.NewEVMBlockContext(ctx, cfg)
	if cfg.BlockOverrides != nil {
		cfg.BlockOverrides.Apply(&blockCtx)
	}
	evm = k.NewEVMWithBlockCtx(ctx, msg, cfg, stateDB, blockCtx)
```

yihuang replied:
we can even put the state/block overrides logic into the `NewEVM` function, since it has all the required information through the `cfg` paramter.

yihuang replied:
https://github.com/ethereum/go-ethereum/pull/26414
FYI, they have block overrides in eth_call as well

mmsqe replied:
also saw that, but need bump to 1.12.x

yihuang replied:
Do we need their changes? we can simply add the parameter to eth_call api and pass all along to grpc handler and ApplyMessage

mmsqe committed:
Update x/evm/keeper/state_transition.go

Co-authored-by: yihuang <huang@crypto.com>
Signed-off-by: mmsqe <mavis@crypto.com>

mmsqe committed:
Apply suggestions from code review

mmsqe committed:
fix noop trace check

this bug was introduced in https://github.com/evmos/ethermint/commit/d4621f3e822f202067c34d33ebceb6eaae5d803c

Commit:
Merge branch 'develop' into block_overrides
